ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -136,50 +140,65 @@ predicate returnStep(Node nodeFrom, Node nodeTo) {
  *    z = x.content
  * end
  * ```
- * for the content write `x.content = y`, we will have `content` being the
+ * for the content write `x.content = y`, we will have `contents` being the
  * literal string `"content"`, `nodeFrom` will be `y`, and `nodeTo` will be the
  * `Foo` object created on the first line of the function. This means we will
  * track the fact that `x.content` can have the type of `y` into the assignment
  * to `z` inside `bar`, even though this content write happens _after_ `bar` is
  * called.
  */
-predicate basicStoreStep(Node nodeFrom, Node nodeTo, string content) {
+predicate basicStoreStep(Node nodeFrom, Node nodeTo, TypeTrackerContentSet contents) {
+  postUpdateStoreStep(nodeFrom, nodeTo, contents)
+  or
+  exists(
```

Comment:
You should be able to use `summarySetterStep` from `FlowSummaryImpl` instead.

Reply:
That gives us a step inside the synthesised callable, whereas we need a step at the call site.

If we start doing bottom-up summarisation of type-tracking steps then we could use `summarySetterStep` instead. However, API graphs would still only want the step at the original call site so the logic is still needed somewhere.

Reply:
`summarySetterStep` is a step at the call site.

Reply:
Ah, sorry I was too a bit quick to conclude what was going on in that predicate.

It does seem to include taint steps though, which we don't want here.

Would you prefer if I add `boolean valuePreserving` as a parameter to that predicate so we can use it from here?

Reply:
👍

Reply:
Unfortunately this leads to non-monotonic recursion, since that predicate ultimately depends on `SummarizedCallable.getACall` whereas we need something that only depends on `getACallSimple`.

Reply:
Ah, right.